### PR TITLE
Redesign Notification Framework

### PR DESCRIPTION
**Pull request by xzhflying**
How to store data
==============
**Activity is a kind of record to note who did what. Notifications are based on activities. Since there are some activities then we need to notify users what happened. So the activities should have many kinds of notification styles**

activities table
-------------------
This table is provided by public_activity gem which will mainly store:
* owner:   We regard this field as the sender of this activity. Most of time should be course
* trackable:   The object that is involved in this activity. (e.g. Achievement, Level ...)
* recipient:   The receiver of this activity. Most of time should be User object.
* key:   The key which indicates the path of template

notification_styles table
------------------------
* activity:  References activities table
* notification_type: The style used to show the notification

What if we just want to send a popup or any other notification types without an existing activity record? We can create a fake activity as the type of :notify. So that we can also store our data. And this kind of activity( :notify) will not be loaded when we want to load recent activities in course show page.

To create a notification:
===================
First you should add `acts_as_notifiable` in the **Model** that you want to send out notifications or activity feeds.( e.g. Level, achievement...)
Use API: `object.notify(recipient, sender, types: {})`. Object stands for the instance of your model
For example, if you want to create three types of notifications based on a achievement object for a user in a course:
`object.notify(user, course, types: { activity: :ActivityType, popup: true, email: true } )`

This notification is for a given user, so the recipient should be the user. Usually, most notifications are sent out by the course like (achievement, level, announcement). So the course should be the sender of this notification. 

`popup: , email: , activity:` are all optional. You can create any combinations of notifications among them. But it certainly doesn't make sense if you do not input any type.

popup: 
---------------
Could be set as enable by `popup: true`. After that You should provide styles in **view/ModelName/notifications/ActivityType/_popup** file. Activity type here will be the type that you input in the API's `activity:` field. If you did not filling this field, the Activity type will be **notify**

email: 
--------
Could be set as enable by `email: true`. After that, you should provide email style in **view/ModelName/notifications/ActivityType/_email** file. In that file you may want to use `message.subject =` to set your email's subject. Meanwhile your email will also use **view/layouts/notifications/email.html.slim** as its layout.

activity:
----------
Could take in a **symbol** as the type name of activity. ( e.g. If one object is created in some controller, you may need to create an activity as `:created`) Activities are shown in course's show view. But you need to provide templates for each kind of activity type you have generated in **view/ModelName/notifications/ActivityType/_activity_feed** file. 

For example, if someone reached some level. You may want to create an :updated activity for level object. To show this activity, you should provide the template. In that template, you may want to write:
        `I18n.t('reached') a.trackable.level_number`  (you can use a.trackable to get your object record in your template)

### CODE REVIEW

File: db/migrate/20150511134159_create_notifications.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+class CreateNotifications < ActiveRecord::Migration
```
Comment (lowjoel) [inline db/migrate/20150511134159_create_notifications.rb]: Squash your migration down to one file.

File: app/models/popup_notification.rb [hunk 2]
```diff
@@ -0,0 +1,4 @@
+class PopupNotification < ActiveRecord::Base
```
Comment (lowjoel) [inline app/models/popup_notification.rb]: Either make this Notification::Popup (which I think was what you were asking...) or make this `NotificationPopup` On hindsight  I think Notification::Popup is better.

File: db/migrate/20150511140404_create_popup_notifications.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+class CreatePopupNotifications < ActiveRecord::Migration
+  def change
+    create_table :popup_notifications do |t|
+      t.string :position
+      t.string :title
+      t.text :content
+      t.string :button_text
+      t.string :image_url
+      t.string :link
+      t.boolean :sharable
```
Comment (lowjoel) [inline db/migrate/20150511140404_create_popup_notifications.rb]: what us thus for?

File: db/migrate/20150511140404_create_popup_notifications.rb [hunk 4]
```diff
@@ -0,0 +1,15 @@
+class CreatePopupNotifications < ActiveRecord::Migration
+  def change
+    create_table :popup_notifications do |t|
+      t.string :position
```
Comment (lowjoel) [inline db/migrate/20150511140404_create_popup_notifications.rb]: Why position?

File: db/migrate/20150511140404_create_popup_notifications.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+      t.string :title
```
Comment (lowjoel) [inline db/migrate/20150511140404_create_popup_notifications.rb]: I got a feeling maybe we should use views to show this. Please look at the `public_activity` gem to see how they handle notifications. Storing translatable strings in the database is a bad idea.

**Comment by lowjoel** [comment]
What about #159?

File: db/migrate/20150511140404_create_popup_notifications.rb (hunk 4 above)
Comment (xzhflying) [inline db/migrate/20150511140404_create_popup_notifications.rb]: cuz, we may have central popup or right side popup. Use position to determine which kind of popup it is.

File: app/models/popup_notification.rb (hunk 2 above)
Comment (xzhflying) [inline app/models/popup_notification.rb]: ok, it is just what I wanted to ask :-p

File: db/migrate/20150511140404_create_popup_notifications.rb (hunk 3 above)
Comment (xzhflying) [inline db/migrate/20150511140404_create_popup_notifications.rb]: It is used to determine if this popup should provide a button(or something else) which could handle sharing to FB(or something else)

File: db/migrate/20150515041839_create_activities.rb [hunk 6]
```diff
@@ -0,0 +1,23 @@
+# Migration responsible for creating a table with activities
+class CreateActivities < ActiveRecord::Migration
+  # Create table
+  def self.up
```
Comment (allenwq) [inline db/migrate/20150515041839_create_activities.rb]: @lowjoel How shall we deal with migration method has too many lines  ?

Comment (xzhflying) [inline db/migrate/20150515041839_create_activities.rb]: LOL

Comment (lowjoel) [inline db/migrate/20150515041839_create_activities.rb]: migration can ignore length. I set it in the .rubocop.yml file but Hound ignores the file.

Comment (allenwq) [inline db/migrate/20150515041839_create_activities.rb]: I see.

**Comment by xzhflying** [comment]
@lowjoel @allenwq Except a 'show_popups' helper which used to render popups' templates, the rest has been completed. The main API is Notification.notify

**Comment by allenwq** [comment]
@lowjoel I guess coverall breaks ?

File: app/controllers/course/controller.rb [hunk 7]
```diff
@@ -63,4 +64,10 @@ def current_component_host
     @current_component_host ||= Course::ComponentHost.new(current_tenant.settings,
                                                           current_course.settings)
   end
+
+  def load_unread_popups
+    return unless current_user && current_course
+    @popups = Notification.unread_by(current_user).joins(:activity).
```
Comment (allenwq) [inline app/controllers/course/controller.rb]: Why Notification.unread_by returns popups ?

File: app/models/notification.rb [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+  def self.notify(user, course, object, options = {})
+    include_public_activity(object)
+    unless options.key?(:activity)
+      notify_only(user, course, object, options)
+      return
+    end
+    activity_and_notify(user, course, object, options)
+  end
+
+  def self.include_public_activity(object)
```
Comment (allenwq) [inline app/models/notification.rb]: You should make all the private methods private

File: app/controllers/course/controller.rb [hunk 9]
```diff
@@ -63,4 +64,16 @@ def current_component_host
     @current_component_host ||= Course::ComponentHost.new(current_tenant.settings,
                                                           current_course.settings)
   end
+
+  def recent_activities(number)
```
Comment (lowjoel) [inline app/controllers/course/controller.rb]: Move this to a concern.

File: app/controllers/course/courses_controller.rb [hunk 10]
```diff
@@ -1,5 +1,6 @@
 class Course::CoursesController < Course::Controller
   def show #:nodoc:
+    @activities = recent_activities(20)
```
Comment (lowjoel) [inline app/controllers/course/courses_controller.rb]: Why is this needed? All controllers with a course loaded would need this right?

File: app/mailers/notification_mailer.rb [hunk 11]
```diff
@@ -0,0 +1,12 @@
+class NotificationMailer < ApplicationMailer
+  layout 'notification_email'
```
Comment (lowjoel) [inline app/mailers/notification_mailer.rb]: Can we use a `/` to delimit?

File: app/models/notification.rb [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+      notify_only(user, course, object, options)
+      return
+    end
+    activity_and_notify(user, course, object, options)
+  end
+
+  class << self
+    private
+
+    def include_public_activity(object)
```
Comment (lowjoel) [inline app/models/notification.rb]: Can't you include this into ActiveRecord::Base in an initialiser?

File: app/models/notification.rb [hunk 13]
```diff
@@ -0,0 +1,55 @@
+class Notification < ActiveRecord::Base
+  belongs_to :activity, inverse_of: :notifications, dependent: :destroy
+
+  acts_as_readable on: :updated_at
+
+  def self.notify(user, course, object, options = {})
+    include_public_activity(object)
+    unless options.key?(:activity)
```
Comment (lowjoel) [inline app/models/notification.rb]: I don't understand why you need to have activity/notify/both

File: app/models/notification.rb [hunk 14]
```diff
@@ -0,0 +1,55 @@
+class Notification < ActiveRecord::Base
+  belongs_to :activity, inverse_of: :notifications, dependent: :destroy
+
+  acts_as_readable on: :updated_at
+
+  def self.notify(user, course, object, options = {})
```
Comment (lowjoel) [inline app/models/notification.rb]: Use Keyword arguments.

Comment (lowjoel) [inline app/models/notification.rb]: Also, what if I do not want an email notification?

What are all the options you support? You did not document.

File: app/models/notification.rb [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+    end
+
+    def popup_notify(activity, popup_type)
+      popup = Notification.new(activity_id: activity.id, notification_type: :popup,
+                               style: popup_type)
+      popup.save
+    end
+
+    def email_notify(user, course, object, options)
+      template = '/notifications/email_notifications/' + options[:email].to_s
```
Comment (lowjoel) [inline app/models/notification.rb]: Use a string injection

**Comment by lowjoel** [comment]
Design questions:

> could take in an array, and the types(symbols) in array will be the templates' names which are used to show those popups. That means you should provide those templates in view/notifications/popups/ directory. (eg. you may want to create a popup named level_up)
Why give the templates names? The templates should follow key of the activity. Same with all the other types of notifications.

> And of course, if you want to send an email notification, you should input an :email_subject as a parameter in notify function.
This should be a `content_for` block in the layout of the email.

> could only take in one of these four symbols: :create, :read, :update, :destroy. These stand for CRUD methods. If one object is created in some controller, you may need to create an activity as :create. Activities are shown in course's show view. But you need to provide templates for each kind of activity(CRUD) you have generated.
Where? Is the API of your framework consistent? Why do you specify paths in one API, but use convention in another? Rails advocates convention over configuration.

I also do not think that the notification need to be restricted to just these 4 symbols.

File: app/models/notification.rb [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+    private
+
+    def include_public_activity(object)
+      return if object.class.include?(PublicActivity::Common)
+      object.class.include(PublicActivity::Common)
+    end
+
+    def notify_only(user, course, object, options)
+      if options.key?(:popup)
+        activity = object.create_activity :notification, owner: course, recipient: user
```
Comment (lowjoel) [inline app/models/notification.rb]: If the method is called notify_only, why is there an activity created?

File: app/views/layouts/notification_email.html.slim [hunk 17]
```diff
@@ -0,0 +1,10 @@
+h4
+  = I18n.t('notification_mailer.notification.greeting', user: @user.name)
```
Comment (lowjoel) [inline app/views/layouts/notification_email.html.slim]: current_user

File: app/views/layouts/notification_email.html.slim [hunk 18]
```diff
@@ -0,0 +1,10 @@
...
+
+= yield
+
+h4
+  = I18n.t('notification_mailer.notification.leave_taking')
```
Comment (lowjoel) [inline app/views/layouts/notification_email.html.slim]: what's this?

File: app/views/layouts/notification_email.html.slim [hunk 19]
```diff
@@ -0,0 +1,10 @@
...
+
+h4
+  = I18n.t('notification_mailer.notification.signature')
```
Comment (lowjoel) [inline app/views/layouts/notification_email.html.slim]: `notification.sign_off`

File: app/views/layouts/notification_email.html.slim (hunk 18 above)
Comment (lowjoel) [inline app/views/layouts/notification_email.html.slim]: oh, this is called 'greeting'

File: app/mailers/notification_mailer.rb (hunk 11 above)
Comment (xzhflying) [inline app/mailers/notification_mailer.rb]: means `notification/email` ?

File: app/models/notification.rb (hunk 13 above)
Comment (xzhflying) [inline app/models/notification.rb]: Will we use activity to send email notification?

File: app/models/notification.rb (hunk 14 above)
Comment (xzhflying) [inline app/models/notification.rb]: I did... At the top of this page, I wrote a description for this pr...

Comment (lowjoel) [inline app/models/notification.rb]: Yah, I saw after reviewing. Still confused after reading it.

File: app/views/layouts/notification_email.html.slim (hunk 17 above)
Comment (xzhflying) [inline app/views/layouts/notification_email.html.slim]: this `@user` is an instance variable defined in notification mailer

File: app/models/notification.rb (hunk 16 above)
Comment (xzhflying) [inline app/models/notification.rb]: because, I need to store that notification information in activities table. In other words, I store data in activities table, if one activity need to send a notification as well (most of time, an activity need to send a notification too), I store its notification type in notifications table and reference activities table for (user, course, object) information

File: app/controllers/course/courses_controller.rb (hunk 10 above)
Comment (xzhflying) [inline app/controllers/course/courses_controller.rb]: ...not quite understand. I think we just need to load recent activities in course show page... Will we show activities in other pages?

File: app/models/notification.rb (hunk 12 above)
Comment (xzhflying) [inline app/models/notification.rb]: Is this necessary? That will cause every active records have this gem's module...

File: app/models/notification.rb (hunk 15 above)
Comment (xzhflying) [inline app/models/notification.rb]: ok~

**Comment by xzhflying** [comment]
@lowjoel modify some doc and codes. Maybe some of my descriptions are misleading, but I'm willing to answer the questions that may make you confused. And correct them.

File: app/controllers/concerns/application_activities_concern.rb [hunk 20]
```diff
@@ -0,0 +1,9 @@
+module ApplicationActivitiesConcern
+  extend ActiveSupport::Concern
+
+  def recent_activities(course, number)
+    return if number < 0 || !number.is_a?(Integer)
```
Comment (lowjoel) [inline app/controllers/concerns/application_activities_concern.rb]: Don't check number to be an integer.

File: app/controllers/concerns/application_activities_concern.rb [hunk 21]
```diff
@@ -0,0 +1,9 @@
...
+    PublicActivity::Activity.where(owner: course).where.not('key like ?', '%notification').
```
Comment (lowjoel) [inline app/controllers/concerns/application_activities_concern.rb]: So you have three fields in a notification:

 - Trackable
 - Owner
 - Recipient

How do you know which field is used for what? How is your design consistent?

File: app/controllers/concerns/application_activities_concern.rb [hunk 22]
```diff
@@ -0,0 +1,9 @@
...
+      order('created_at desc').includes(:trackable, :recipient).first(number)
```
Comment (lowjoel) [inline app/controllers/concerns/application_activities_concern.rb]: If I don't specify how many I want?

File: app/controllers/course/controller.rb [hunk 23]
```diff
@@ -63,4 +64,10 @@ def current_component_host
     @current_component_host ||= Course::ComponentHost.new(current_tenant.settings,
                                                           current_course.settings)
   end
+
+  def load_unread_popups
```
Comment (lowjoel) [inline app/controllers/course/controller.rb]: Place in a concern.

File: app/controllers/course/courses_controller.rb [hunk 24]
```diff
@@ -1,5 +1,6 @@
 class Course::CoursesController < Course::Controller
   def show #:nodoc:
+    @activities = recent_activities(current_course, 20)
```
Comment (lowjoel) [inline app/controllers/course/courses_controller.rb]: I'm still not convinced how this works or what this is for.

File: app/mailers/notification_mailer.rb [hunk 25]
```diff
@@ -0,0 +1,12 @@
+class NotificationMailer < ApplicationMailer
+  layout 'notifications/email'
+
+  def notify(user, course, template_url, subject, obj)
+    @user = user
+    @course = course
+    @obj = obj
+    mail(to: user.email, subject: subject) do |format|
```
Comment (lowjoel) [inline app/mailers/notification_mailer.rb]: Why can't the view set the subject?

Comment (lowjoel) [inline app/mailers/notification_mailer.rb]: Like how templates set the title?

File: app/models/notification.rb [hunk 26]
```diff
@@ -0,0 +1,55 @@
+class Notification < ActiveRecord::Base
+  belongs_to :activity, inverse_of: :notifications, dependent: :destroy
+
+  acts_as_readable on: :updated_at
+
+  def self.notify(user, course, object, types = {})
```
Comment (lowjoel) [inline app/models/notification.rb]: Document your API calls. I won't be reading your PR to see how to use this method.

File: app/models/notification.rb (hunk 12 above)
Comment (lowjoel) [inline app/models/notification.rb]: isn't that the point? You're injecting the module into every class anyway. When you have eager loading in production this method consumes even more memory.

Comment (lowjoel) [inline app/models/notification.rb]: And I don't think this works with inherited models.

File: app/models/notification.rb (hunk 26 above)
Comment (lowjoel) [inline app/models/notification.rb]: Use keyword arguments - I said so in the last review?

File: app/models/notification.rb [hunk 27]
```diff
@@ -0,0 +1,55 @@
+class Notification < ActiveRecord::Base
```
Comment (lowjoel) [inline app/models/notification.rb]: The design of your notification class is incredibly awkward:

 - methods are named X_and_Y, and everything sends an email anyway. Shouldn't the email be optional?
 - All the views are are dumped in a common notifications directory. How does this design scale for a large application with multiple components like Coursemology?
 - How do you know your design will work for all the existing notifications that Coursemology v1 would require to make?

File: app/views/layouts/notifications/email.html.slim [hunk 28]
```diff
@@ -0,0 +1,10 @@
+h4
+  = I18n.t('notification_mailer.notification.greeting', user: @user.name)
+
+= yield
+
+h4
+  = I18n.t('notification_mailer.notification.leave_taking')
```
Comment (lowjoel) [inline app/views/layouts/notifications/email.html.slim]: I already said about what this should be called. Dig the PR history.

File: app/models/activity.rb [hunk 29]
```diff
@@ -0,0 +1,4 @@
+# Activity model for customisation & custom methods
+class Activity < PublicActivity::Activity
+  has_many :notifications, inverse_of: :activity, dependent: :destroy
```
Comment (lowjoel) [inline app/models/activity.rb]: I still do not understand why one activity can have many notifications. That sounds weird.

Comment (lowjoel) [inline app/models/activity.rb]: I create one notification. How can one activity have many notifications?

File: app/controllers/concerns/application_activities_concern.rb (hunk 21 above)
Comment (xzhflying) [inline app/controllers/concerns/application_activities_concern.rb]: In my opinion, these three fields are just three names. Actually they are the same. Each of them will use polymorphic to store an object. Trackable is for the object you want to track. In our use cases it should be the input object like level, achievement...  As for owner and recipient, I think notification is for a user and generated by course or something else. So I use recipient filed to store user object. If we will have some expending, maybe we just need to change owner filed.

File: app/controllers/concerns/application_activities_concern.rb (hunk 22 above)
Comment (xzhflying) [inline app/controllers/concerns/application_activities_concern.rb]: To load every activities?...

File: app/controllers/course/controller.rb (hunk 23 above)
Comment (xzhflying) [inline app/controllers/course/controller.rb]: ok I forgot yesterday...

File: app/controllers/course/courses_controller.rb (hunk 24 above)
Comment (xzhflying) [inline app/controllers/course/courses_controller.rb]: maybe function name is misleading... It should be `load_recent_activities` and what I thought is that we just need to load recent activities in course's show page.

File: app/models/notification.rb (hunk 26 above)
Comment (xzhflying) [inline app/models/notification.rb]: Document means google doc?

File: app/models/activity.rb (hunk 29 above)
Comment (xzhflying) [inline app/models/activity.rb]: umm... maybe I can cancel this relation.

File: app/views/layouts/notifications/email.html.slim (hunk 28 above)
Comment (xzhflying) [inline app/views/layouts/notifications/email.html.slim]: umm... greeting? But it already exits...

File: app/models/notification.rb (hunk 27 above)
Comment (xzhflying) [inline app/models/notification.rb]: * email is optional, I just divide notify into 'with a given activity record' and 'without a given activity record'. I will check if types have an email key words in these two situation.
* Where am I supposed to put those in?   **view/activities**, **view/popups** and **view/notification_mailer**?
* Existing notifications? Means `popup_level` 'popup_achievement' and 'activities shown in course page'?

File: app/controllers/concerns/application_activities_concern.rb (hunk 21 above)
Comment (lowjoel) [inline app/controllers/concerns/application_activities_concern.rb]: If they are the same then what is the point of a database schema? You cannot shove values in willy-nilly.

I agree with your definition of trackable.

I am not convinced your decision to assign owners and recipients is the best available. How do you reconcile things like activity feeds with popups?

File: app/controllers/concerns/application_activities_concern.rb (hunk 22 above)
Comment (lowjoel) [inline app/controllers/concerns/application_activities_concern.rb]: Your method doesn't allow that.

File: app/controllers/course/courses_controller.rb (hunk 24 above)
Comment (lowjoel) [inline app/controllers/course/courses_controller.rb]: Then why does every course controller need to have this method included?

File: app/models/activity.rb (hunk 29 above)
Comment (lowjoel) [inline app/models/activity.rb]: > activity is a kind of notification

does that sound correct? That implies that all activities are notifications. Please figure out the cardinality of the associations correctly. Associations are not just thrown into the application.

File: app/models/notification.rb (hunk 26 above)
Comment (lowjoel) [inline app/models/notification.rb]: We use YARD.

File: app/models/notification.rb (hunk 27 above)
Comment (lowjoel) [inline app/models/notification.rb]: Your methods do not show proper levels of abstraction. It is confusing for me to read your code. Please refactor.

The views should be placed locally within the component. If I'm doing announcements I should be placing the notification views in the announcements view directory.

Existing notifications refer to the notifications that v1 currently sends. How do you map v1 notifications to your framework? Have you considered that?

File: app/views/layouts/notifications/email.html.slim (hunk 28 above)
Comment (lowjoel) [inline app/views/layouts/notifications/email.html.slim]: https://www.nmu.edu/writingcenter/node/194

"Complimentary Close"

File: app/models/activity.rb (hunk 29 above)
Comment (xzhflying) [inline app/models/activity.rb]: .... I said do not think...  Means activity is not a notification. It is just a record... @lowjoel  But I can remove that relation..

File: app/models/concerns/notify_concern.rb [hunk 30]
```diff
@@ -0,0 +1,70 @@
+module NotifyConcern
```
Comment (allenwq) [inline app/models/concerns/notify_concern.rb]: @xzhflying have you documented how to use your framework ? or examples ?

If I understand correctly, this concern is supposed to included by models, then they can send notifications  ?

File: app/models/activity.rb [hunk 31]
```diff
@@ -0,0 +1,15 @@
...
+
+  # Translate the key of activity to path
+  #   The path will be /ClassName/notifications/activity_type/activity_feed
+  #
+  # @return [String]
+  def key_to_path
```
Comment (allenwq) [inline app/models/activity.rb]: how about <code>template_path</code>, <code>key_to_path</code> is a bit hard to understand

**Comment by xzhflying** [comment]
@lowjoel @allenwq

File: app/models/concerns/notify_concern.rb (hunk 30 above)
Comment (allenwq) [inline app/models/concerns/notify_concern.rb]: Wrap this inside a function like <code>acts_as_notifiable</code>, so others can include you by:

```ruby
class Achievement
  acts_as_notifiable  # instead of include
end
```

File: app/models/concerns/notify_concern.rb [hunk 32]
```diff
@@ -0,0 +1,70 @@
...
+
+  # Send notifications according to input types based on object which calls the method.
+  #
+  # @param [Object] recipient who does these notifications send for
+  # @param [Object] generator who generate this notification
+  # @param [Hash] types which kinds of notifications you want to send
+  # @option types [Symbol] :activity The type of activity
+  # @option types [True] :popup Enable to send a popup notification
+  # @option types [True] :email Enable to send an email notification
+  def notify(recipient, generator, types: {})
```
Comment (allenwq) [inline app/models/concerns/notify_concern.rb]: What is <code>generator</code> ? <code> sender</code> ?

File: app/models/concerns/notify_concern.rb [hunk 33]
```diff
@@ -0,0 +1,70 @@
...
+      "#{self.class.name.underscore.pluralize}.notifications.notify"
+    end
+  end
+
+  # Get email notification's template path from the key of activity which this notification is
+  #   based on
+  #
+  # @param [Symbol] activity The type of activity
+  # @return [String]
+  def get_email_template(key)
```
Comment (allenwq) [inline app/models/concerns/notify_concern.rb]: no <code>get</code> or <code>set</code> in ruby

File: app/models/activity.rb (hunk 29 above)
Comment (allenwq) [inline app/models/activity.rb]: Is it possible for you to remove this association or let notification to have activity ?

File: app/models/concerns/notify_concern.rb (hunk 32 above)
Comment (xzhflying) [inline app/models/concerns/notify_concern.rb]: yes... better to use `sender`?

File: app/models/concerns/notify_concern.rb (hunk 33 above)
Comment (xzhflying) [inline app/models/concerns/notify_concern.rb]: alright, my bad...

File: app/models/activity.rb (hunk 31 above)
Comment (xzhflying) [inline app/models/activity.rb]: ok...

File: app/models/concerns/notify_concern.rb (hunk 30 above)
Comment (xzhflying) [inline app/models/concerns/notify_concern.rb]: got it

Comment (lowjoel) [inline app/models/concerns/notify_concern.rb]: Wait. I said *NOT* to require classes to explicitly include this concern.

File: app/models/activity.rb [hunk 34]
```diff
@@ -0,0 +1,15 @@
+# Activity model for customisation & custom methods
+class Activity < PublicActivity::Activity
+  has_many :notification_styles, inverse_of: :activity, dependent: :destroy
```
Comment (lowjoel) [inline app/models/activity.rb]: I still do not get your design.

Comment (xzhflying) [inline app/models/activity.rb]: one Activity record has many kinds of methods( way ) to notify users what happened. I call these methods `notification_styles`. It is better than calling them `notifications`.

File: app/models/concerns/notify_concern.rb (hunk 30 above)
Comment (xzhflying) [inline app/models/concerns/notify_concern.rb]: ...alright,  easy to change

File: app/controllers/concerns/course/activities_concern.rb [hunk 35]
```diff
@@ -0,0 +1,20 @@
...
+  extend ActiveSupport::Concern
+
+  # Load recent activity feeds
+  #
+  # @param [Course] course course which activities are related to
+  # @param [Hash]
+  # @option [Integer] :number The number of activities that will be loaded
+  #
+  # @return [Array<Activity>]
+  def recent_activities(course, number: false)
```
Comment (lowjoel) [inline app/controllers/concerns/course/activities_concern.rb]: If this is only ever going to be referenced in the Course controller, why not put it there?

Comment (lowjoel) [inline app/controllers/concerns/course/activities_concern.rb]: and `number: false` is really weird. `number: nil`?

File: app/controllers/concerns/course/activities_concern.rb [hunk 36]
```diff
@@ -0,0 +1,20 @@
...
+    if number
```
Comment (lowjoel) [inline app/controllers/concerns/course/activities_concern.rb]: You can chain this.
```ruby
result = Activity.where(owner: course).where.not('key like ?', '%notify').
  order('created_at desc').includes(:trackable, :recipient)
result = result.first(number) unless number.nil?
```

Comment (lowjoel) [inline app/controllers/concerns/course/activities_concern.rb]: Since you have a non-trivial query here you should be using squeel.

File: app/controllers/concerns/course/activities_concern.rb [hunk 37]
```diff
@@ -0,0 +1,20 @@
+module Course::ActivitiesConcern
+  extend ActiveSupport::Concern
+
+  # Load recent activity feeds
+  #
+  # @param [Course] course course which activities are related to
+  # @param [Hash]
```
Comment (lowjoel) [inline app/controllers/concerns/course/activities_concern.rb]: ..what?

File: app/controllers/concerns/course/activities_concern.rb [hunk 38]
```diff
@@ -0,0 +1,20 @@
...
+  # @option [Integer] :number The number of activities that will be loaded
```
Comment (lowjoel) [inline app/controllers/concerns/course/activities_concern.rb]: Should be `@param` since it's now keyword arguments.

File: app/controllers/concerns/course/notifications_concern.rb [hunk 39]
```diff
@@ -0,0 +1,15 @@
+module Course::NotificationsConcern
+  extend ActiveSupport::Concern
+
+  included do
+    before_action :load_unread_popups
+  end
+
+  private
+
+  def load_unread_popups #:nodoc:
```
Comment (lowjoel) [inline app/controllers/concerns/course/notifications_concern.rb]: does it matter if i only load the current course? What if it is from another course? Do I want it to appear?

File: app/models/activity.rb [hunk 40]
```diff
@@ -0,0 +1,15 @@
...
+
+  # Translate the key of activity to path
+  #   The path will be /ClassName/notifications/activity_type/activity_feed
+  #
+  # @return [String]
+  def template_path
+    partial_root = ''
+    path = key.split('.')
+    path.unshift partial_root
```
Comment (lowjoel) [inline app/models/activity.rb]: `path.unshift ''`?

Comment (lowjoel) [inline app/models/activity.rb]: format('/%s/activity_feed', key.replace('.', '/'))?

File: app/models/concerns/notify_concern.rb (hunk 30 above)
Comment (lowjoel) [inline app/models/concerns/notify_concern.rb]: I discussed with @allenwq, maybe just directly create the activity? No need to include the common module.

File: app/models/concerns/notify_concern.rb [hunk 41]
```diff
@@ -0,0 +1,70 @@
...
+      "#{self.class.name.underscore.pluralize}.notifications.notify"
+    end
+  end
+
+  # Get email notification's template path from the key of activity which this notification is
+  #   based on
+  #
+  # @param [Symbol] activity The type of activity
+  # @return [String]
+  def email_template(key)
```
Comment (lowjoel) [inline app/models/concerns/notify_concern.rb]: I've seen this motif many places. Please factor out.

File: app/models/activity.rb (hunk 40 above)
Comment (xzhflying) [inline app/models/activity.rb]: Got it

File: app/controllers/concerns/course/activities_concern.rb (hunk 38 above)
Comment (xzhflying) [inline app/controllers/concerns/course/activities_concern.rb]: has been modified in new version

File: app/controllers/concerns/course/activities_concern.rb (hunk 35 above)
Comment (xzhflying) [inline app/controllers/concerns/course/activities_concern.rb]: What if there are some new requirements that need to show activity feeds in other places...
